### PR TITLE
Fix for updateStatus segfault 

### PR DESCRIPTION
**Pull request by abake48**
Related to #5

In the event that `readResponse` is empty, a call to `updateStatus` will still try to access the empty vector, leading to a segfault.

### CODE REVIEW

**Comment by MarqRazz** [comment]
Can we add a better way of tracking that the [gripper failed to connect](https://github.com/PickNikRobotics/ros2_robotiq_gripper/blob/55e5525413bc68ee5ae091b2ebfed5b915af79f5/robotiq_driver/src/robotiq_gripper_interface.cpp#L81-L84)?

One suggestion would be to actually implement the [on_init](https://github.com/PickNikRobotics/ros2_robotiq_gripper/blob/55e5525413bc68ee5ae091b2ebfed5b915af79f5/robotiq_driver/include/robotiq_driver/hardware_interface.hpp#L56) method and `return CallbackReturn::ERROR;` if we were not able to establish communication with the gripper (and print a useful error for the operator).

**Comment by abake48** [comment]
It looks like the [on_activate](https://github.com/PickNikRobotics/ros2_robotiq_gripper/blob/ad50ff993f24febd945af14d03991787375c26f5/robotiq_driver/src/hardware_interface.cpp#L161) method will return a `CallbackReturn::ERROR` when `activateGripper` returns `False`. I added a logging statement to the failure condition for better traceability of the error and an error status so it doesn't get stuck in `activateGripper` in the event an empty response is received while activating the gripper is in progress.

File: robotiq_driver/CMakeLists.txt [hunk 1]
```diff
@@ -5,6 +5,8 @@ if(CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")
   add_compile_options(-Wall -Wextra -Wpedantic)
 endif()
 
+
+
```
Comment (sea-bass) [inline robotiq_driver/CMakeLists.txt:9]: remove?

File: robotiq_driver/include/robotiq_driver/robotiq_gripper_interface.hpp [hunk 2]
```diff
@@ -46,14 +46,14 @@ class RobotiqGripperInterface
   /**
    * @brief Activates the gripper.
    *
-   * @return true The gripper was activated successfully.
-   * @return false The gripper was not activated.
+   * @throw std::runtime_error on failure to successfully communicate with gripper port
```
Comment (sea-bass) [inline robotiq_driver/include/robotiq_driver/robotiq_gripper_interface.hpp:49]: Add full stop at end of this line, same w the others in this header

File: robotiq_driver/src/gripper_interface_test.cpp [hunk 3]
```diff
@@ -35,58 +35,61 @@ constexpr auto kComPort = "/dev/ttyUSB0";
 constexpr auto kSlaveID = 0x09;
 
 int main() {
-    RobotiqGripperInterface gripper(kComPort, kSlaveID);
+    try{
```
Comment (sea-bass) [inline robotiq_driver/src/gripper_interface_test.cpp:38]:
```suggestion
    try {
```

Same with all the other try blocks in this PR

File: robotiq_driver/src/gripper_interface_test.cpp [hunk 4]
```diff
@@ -35,58 +35,61 @@ constexpr auto kComPort = "/dev/ttyUSB0";
...
+        RobotiqGripperInterface gripper(kComPort, kSlaveID);
```
Comment (sea-bass) [inline robotiq_driver/src/gripper_interface_test.cpp:39]: in the year of our Lord 2023!

File: robotiq_driver/src/gripper_interface_test.cpp [hunk 5]
```diff
@@ -35,58 +35,61 @@ constexpr auto kComPort = "/dev/ttyUSB0";
...
-    std::cout << "Opening gripper...\n";
-    gripper.setGripperPosition(0x00);
-    while (gripper.gripperIsMoving()) {
-        std::this_thread::sleep_for(std::chrono::milliseconds(500));
+        std::cout << "Opening gripper...\n";
+        gripper.setGripperPosition(0x00);
+        while (gripper.gripperIsMoving()) {
+            std::this_thread::sleep_for(std::chrono::milliseconds(500));
+        }
+    } catch(const std::runtime_error &e){
```
Comment (sea-bass) [inline robotiq_driver/src/gripper_interface_test.cpp:90]: same here, apply to whole PR

```suggestion
    } catch (const std::runtime_error &e) {
```

File: robotiq_driver/src/robotiq_gripper_interface.cpp [hunk 6]
```diff
@@ -80,40 +81,49 @@ RobotiqGripperInterface::RobotiqGripperInterface(const std::string& com_port, ui
...
+    throw std::runtime_error("Failed to open gripper port.");
   }
 }
 
-bool RobotiqGripperInterface::activateGripper()
+void RobotiqGripperInterface::activateGripper()
 {
-  auto cmd = createWriteCommand(kActionRequestRegister, { 0x0100, 0x0000, 0x0000 }  // set rACT to 1, clear all
+  const auto cmd = createWriteCommand(kActionRequestRegister, { 0x0100, 0x0000, 0x0000 }  // set rACT to 1, clear all
                                                                                     // other registers.
```
Comment (sea-bass) [inline robotiq_driver/src/robotiq_gripper_interface.cpp:91]: Make sure the comments are aligned b/w these 2 lines

File: robotiq_driver/src/robotiq_gripper_interface.cpp [hunk 7]
```diff
@@ -80,40 +81,49 @@ RobotiqGripperInterface::RobotiqGripperInterface(const std::string& com_port, ui
...
 
-  updateStatus();
-  while (gripper_status_ == GripperStatus::IN_PROGRESS)
-  {
-    std::this_thread::sleep_for(std::chrono::milliseconds(1000));
     updateStatus();
+    
+    if(gripper_status_ == GripperStatus::COMPLETED){
+      return;
+    }
```
Comment (sea-bass) [inline robotiq_driver/src/robotiq_gripper_interface.cpp:102]: same spacing deal

```suggestion
    if (gripper_status_ == GripperStatus::COMPLETED) {
      return;
    }
```

File: robotiq_driver/src/robotiq_gripper_interface.cpp [hunk 8]
```diff
@@ -80,40 +81,49 @@ RobotiqGripperInterface::RobotiqGripperInterface(const std::string& com_port, ui
...
+
+    while (gripper_status_ == GripperStatus::IN_PROGRESS)
+    {
+      std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+      updateStatus();
+    }
```
Comment (sea-bass) [inline robotiq_driver/src/robotiq_gripper_interface.cpp:108]: Interestingly enough, this is now the only bit of code where the opening brace `{` is on a new line. No big deal, but maybe for consistency bring it up to the previous line?

File: robotiq_driver/src/robotiq_gripper_interface.cpp [hunk 9]
```diff
@@ -202,72 +229,75 @@ std::vector<uint8_t> RobotiqGripperInterface::readResponse(size_t num_bytes_requ
 
   if (num_bytes_read != num_bytes_requested)
   {
-    std::cerr << "Requested " << num_bytes_requested << " bytes, but only got " << num_bytes_read << ".\n";
+    throw std::runtime_error("Requested " + std::to_string(num_bytes_requested) 
+                    + " bytes, but only got " + std::to_string(num_bytes_read));
```
Comment (sea-bass) [inline robotiq_driver/src/robotiq_gripper_interface.cpp:233]: not sure what the formatting should be here, but this looks kinda off

File: robotiq_driver/src/robotiq_gripper_interface.cpp [hunk 10]
```diff
@@ -202,72 +229,75 @@ std::vector<uint8_t> RobotiqGripperInterface::readResponse(size_t num_bytes_requ
...
-bool RobotiqGripperInterface::sendCommand(const std::vector<uint8_t>& cmd) {
+void RobotiqGripperInterface::sendCommand(const std::vector<uint8_t>& cmd) {
   size_t num_bytes_written = port_.write(cmd);
   port_.flush();
   if (num_bytes_written != cmd.size())
   {
-    std::cerr << "Attempted to write " << cmd.size() << " bytes, but only wrote " << num_bytes_written << ".\n";
-    return false;
+    throw std::runtime_error("Attempted to write " + std::to_string(cmd.size()) 
+                + " bytes, but only wrote " + std::to_string(num_bytes_written));
```
Comment (sea-bass) [inline robotiq_driver/src/robotiq_gripper_interface.cpp:245]: same here for the alignment formatting

**Comment by sea-bass** [comment]
> Can we add a better way of tracking that the [gripper failed to connect](https://github.com/PickNikRobotics/ros2_robotiq_gripper/blob/55e5525413bc68ee5ae091b2ebfed5b915af79f5/robotiq_driver/src/robotiq_gripper_interface.cpp#L81-L84)?
> 
> One suggestion would be to actually implement the [on_init](https://github.com/PickNikRobotics/ros2_robotiq_gripper/blob/55e5525413bc68ee5ae091b2ebfed5b915af79f5/robotiq_driver/include/robotiq_driver/hardware_interface.hpp#L56) method and `return CallbackReturn::ERROR;` if we were not able to establish communication with the gripper (and print a useful error for the operator). As an example [here is how the UR robot](https://github.com/UniversalRobots/Universal_Robots_ROS2_Driver/blob/main/ur_robot_driver/src/hardware_interface.cpp#L58-L140) handles it.

Tracing this in, the `on_init()` already is in charge of instantiating a `RobotiqGripperInterface` (and @abake48 now catches this exception that the constructor would otherwise throw). So I think we're good in that sense.

**Review by sea-bass — COMMENTED** [review]
Changes look good!

I notice there are two more places in `hardware_interface.cpp` that does not catch exceptions when activating/deactivating the gripper -- so maybe add them here too? 

https://github.com/PickNikRobotics/ros2_robotiq_gripper/blob/835f6a826b89fdfcce671b9e3448f90e09960af9/robotiq_driver/src/hardware_interface.cpp#L164-L170

https://github.com/PickNikRobotics/ros2_robotiq_gripper/blob/835f6a826b89fdfcce671b9e3448f90e09960af9/robotiq_driver/src/hardware_interface.cpp#L220-L221

Besides that, silly pedantic formatting comments.

**Comment by abake48** [comment]
Decided to add clang-format rules because this library didn't have any. The code should be formatted and not rough looking anymore.

**Review by sea-bass — COMMENTED** [review]
Looks good, and thanks for adding clang-format!

I think you missed the 2nd try/catch I pointed out in the `on_deactivate()` method though:

https://github.com/PickNikRobotics/ros2_robotiq_gripper/blob/06b8e48136aafe77873c94059f13d3353012044c/robotiq_driver/src/hardware_interface.cpp#L229-L230

**Review by sea-bass — APPROVED** [review]
LGTM!

Out of curiosity, why the atomic variable addition?

**Comment by abake48** [comment]
Anxiety over variables that cross threads. Probably isn't needed but since it's read and written to in both threads, it made me feel safe to just make that change. Idk if there's a best practice out there for that though.

File: robotiq_driver/src/hardware_interface.cpp [hunk 11]
```diff
@@ -111,10 +111,18 @@ RobotiqGripperHardwareInterface::on_init(const hardware_interface::HardwareInfo&
...
+  try
+  {
+    // Create the interface to the gripper.
+    gripper_interface_ = std::make_unique<RobotiqGripperInterface>(com_port_);
+    gripper_interface_->setSpeed(gripper_speed * 0xFF);
+    gripper_interface_->setForce(gripper_force * 0xFF);
+  }
+  catch (const serial::IOException& e)
+  {
+    RCLCPP_FATAL(kLogger, e.what());
```
Comment (MarqRazz) [inline robotiq_driver/src/hardware_interface.cpp:123]: Can we make this print a better message? All this returns is 
```
[launcher_node-3] [ros2_control_node-1] [RobotiqGripperHardwareInterface 1680906123.713730209]: IO Exception (2): No such file or directory, file /opt/underlay_ws/src/serial/src/impl/unix.cc, line 151.
[launcher_node-3] [ros2_control_node-1] [resource_manager 1680906123.713742143]: Failed to initialize hardware 'RobotiqGripperHardwareInterface'
```

**Review by MarqRazz — APPROVED** [review]
Tested with and without hardware. If we can make it print a better error on failure that would be great!

**Pull request merged**